pg_tmp.1
.Dd February 6, 2025
.Dt PG_TMP 1
.Os
.Sh NAME
.Nm pg_tmp
.Nd quickly spin up temporary PostgreSQL database
.Sh SYNOPSIS
.Nm pg_tmp
.Op Fl kt
.Op Fl d Ar datadir
.Op Fl o Ar pgoptions
.Op Fl p Ar port
.Op Fl w Ar timeout
.Sh DESCRIPTION
The
.Nm
utility creates a temporary instance of PostgreSQL and prints the connection URL
to standard output.
.Nm
spawns a background process that initializes a new database under
.Pa /tmp/ephemeralpg.*/${PGVER}
which is used by subsequent invocations to reduce startup time.
The arguments are as follows:
.Bl -tag -width Ds
.It Fl k
Keep the temporary directory after the server is shut down.
.It Fl t
Use a TCP port selected by
.Xr getsocket 1 .
A port may be selected using the
.Fl p
flag.
Otherwise the path to a Unix socket is returned.
.It Fl d
Specify the temporary directory to use.
May be used with the optional arguments
.Ar initdb ,
.Ar start
and is required for
.Ar stop .
If this option is used
.Nm
will not initialize a new database for subsequent invocations use.
.It Fl o
Specifies options to be passed directly to the
.Xr postgres 1
server.
These options should usually be surrounded by quotes to ensure that they are
passed through as a group.
.It Fl p
TCP port to listen on.
.It Fl w
Shut down and remove the database after the specified
.Ar timeout .
If one or more clients are still connected then
.Nm
sleeps and retries again after the same interval.
A value of 0 will leave the server running indefinitely.
The default is 60 seconds.
.El
.Sh ENVIRONMENT
.Bl -tag -width TMPDIR
.It Ev LC_ALL
Set locale to ensure proper startup on Mac OS.
If not set the default is
.Ql C .
.It Ev TMPDIR
base directory in which to create and run the ephemeral instances of
.Xr postgres 1
.El
.Sh EXAMPLES
Create a temporary database and run a query:
.Bd -literal -offset indent
uri=$(pg_tmp)
psql $uri -f my.sql
.Ed
.Pp
Start a temporary server with a custom extension:
.Bd -literal -offset indent
uri=$(pg_tmp -o "-c shared_preload_libraries=$PWD/auth_hook")
psql $uri -c "SELECT 1"
.Ed
.Pp
Start a temporary database that recycles WAL segments to reduce space
.Bd -literal -offset indent
pg_tmp -o "-c max_wal_size=48MB -c min_wal_size=32M"
.Ed
.Sh SEE ALSO
.Xr pg_ctl 1
.Sh AUTHORS
.An Eric Radman Aq Mt ericshane@eradman.com


NEWS
= Release History

== 3.4: February 6, 2025

 - Set LC_ALL to ensure proper startup on Mac OS
 - Keep server up for any normal client connection
 - Add Gemfile with test dependencies
 - Fix Ubuntu by using $3 for from version string

== 3.3: November 17, 2023

 - Adapt tests to zsh
 - Use ${RUBY} instead of 'ruby' for test runner
 - Avoid using 'set +o posix' for compatibility with Alpine

== 3.2: April 13, 2022

 - flattenjs: exit 0 if the input is empty
 - Drop ddl_compare(1)

== 3.1: October 01, 2020

 - Minor revisions to README and Makefile
 - No longer set 'wal_keep_segments' for compatibility with 13

== 3.0: March 27, 2020

 - Add ddl_compare(1), a utility for creating schema diffs
 - Listen on '*' when the '-t' option is specified

== 2.9: November 21, 2019

 - More portable integration tests
 - Add contrib/flattenjs
 - Detect version based on pg_ctl(1), not psql(1)

== 2.8: September 25, 2019

 - Move hosting from bitbucket.org to github.com
 - Rename tag names from ephemeralpg-X.Y to X.Y

== 2.7: May 22, 2019

 - '-w 0' leaves the server running indefinitely
 - New '-k' flag prevents the temporary directory from being cleaned up
 - Only repeat the last 10 lines of the postgres log if createdb(1) fails
 - Set default check interval to 5 seconds for 'stop'

== 2.6: March 27, 2019

 - Only run getsocket(1) if a port was not specified
 - Reduce disk usage by running checkpoints after 64MB of WAL (default is 1GB)

== 2.5: March 01, 2018

 - Skip over pre-initialized database directories that are not owned by the
   current user
 - Guard against a variant race condition introduced in ephemeralpg-2.4


== 2.4: February 02, 2018

 - Ignore .psqlrc to prevent indeterminate behaviors
 - Guard against race condition if pg_tmp is executed concurrently

== 2.3: August 11, 2017

 - Allow irregular version strings such as "10beta3"
 - Use still-alive query compatible with 10 and 9.x

== 2.2: December 13, 2016

 - Include username in URL when '-t' is specified

== 2.1: October 31, 2016

 - Fix build for getsocket(1) on FreeBSD
 - Add instructions for building getsocket(1) on Solaris

== 2.0: March 29, 2016

 - Allow a temporary directory to be used in combination with the optional
   arguments 'initdb', 'start', and 'stop'
 - 'stop' only sleeps if '-w' was specified

== 1.9: March 09, 2016

 - Unbreak directory cleanup with PG 9.3
 - Turn off synchronous_commit by default
 - Unix sockets are specified using ?host= for compatibility with SQLAlchemy

== 1.8: January 05, 2016

 - Stop a running instance if query for open connections fails
 - Ignore SIGHUP instead of relying on nohup(1)
 - Properly support multiple extra-options to using the flag '-o'

== 1.7: December 12, 2015

 - Properly stop the server that is listening on a TCP port
 - More extensive test framework included: 'make test'

== 1.6: December 10, 2015

 - Exit non-zero if getopt(1) fails
 - Raise an error early if initdb(1) cannot be found
 - Run background initdb and timed stop even if postgres fails to start

== 1.5: December 01, 2015

 - Allow extra-options to be passed directly to postgres(1) on start using '-o'
 - Copy log to STDERR and exit non-zero if server startup fails

== 1.4: October 30, 2015

 - Only remove the temporary directory if postgresql.auto.conf exists
 - Database is started under /tmp/ephemeralpg.*/$PGVER/ instead of
   /tmp/ephemeralpg$PGVER.*/db/
 - More aggressively suppress warnings from nohup(1)

== 1.3: October 23, 2015

 - Display ephemeralpg release number along with usage help text
 - Unrecognized options display usage
 - Temporary directory is named with the major version number
 - Simplify socket URL
 - Drop log file location flag

== 1.2: January 24, 2015

 - Log statements and connect/disconnects.
 - Specify log file location with '-l' flag
 - Database is shutdown only after clients have disconnected. The timeout now
   specifies a check interval
 - Newline is not suppressed if run in a terminal
 - Bugfix in getsock(1) so that only ports above 1024 are returned

== 1.1: January 17, 2015

 - Timeout flag changed to '-w' to match netcat(1)
 - Use '-t' flag to bind to a TCP port using bundled getsocket(1) utility

== 1.0: January 07, 2015

 - Initial release at https://bitbucket.org/eradman/ephemeralpg
 - Cross-platform implementation in less then 100 lines of code



pg_tmp.sh
#!/bin/sh
# pg_tmp
# Quickly spin up a temporary PostgreSQL test databases

usage() {
	echo >&2 "release: ${release}"
	echo >&2 "usage: pg_tmp [-kt] [-d datadir] [-o pgoptions] [-p port] [-w timeout]"
	if [ -z "$1" ]; then
		echo >&2 "hint: use -h to display option summary"
		exit 1
	fi

	cat <<- HELP
		summary:
		    -k            Do not clean up data directory
		    -t            Use TCP instead of local socket
		    -d datadir    Specify directory to initialize
		    -o pgoptions  Run-time parameters for server to use
		    -p port       Port to listen on
		    -w timeout    Delay in cleaning up data directory
		docs:
		    man pg_tmp
	HELP
	exit 1
}

initdb -V > /dev/null || exit 1

trap 'printf "$0: exit code $? on line $LINENO\n" >&2; exit 1' ERR \
	2> /dev/null || exec bash $0 "$@"
trap '' HUP

PGVER=$(pg_ctl -V | awk '{print $3}')
USER_OPTS=""
[ "x$1" == "x-h" ] && usage $1
getopt > /dev/null hktp:w:o:d: "$@" || usage
while [ $# -gt 0 ]; do
	case "$1" in
		-k) KEEP=$1 ;;
		-t) LISTENTO="127.0.0.1" ;;
		-d)
			TD="$2"
			shift
			;;
		-o)
			USER_OPTS="$2"
			shift
			;;
		-p)
			PGPORT="$2"
			shift
			;;
		-w)
			TIMEOUT="$2"
			shift
			;;
		*) CMD=$1 ;;
	esac
	shift
done

[ -n "$LISTENTO" ] && [ -z "$PGPORT" ] && {
	PGPORT="$(getsocket)"
}

case ${CMD:-start} in
	initdb)
		[ -z $TD ] || mkdir -p $TD
		[ -z $TD ] && TD="$(mktemp -d ${SYSTMP:-/tmp}/ephemeralpg.XXXXXX)"
		initdb --nosync -D $TD/$PGVER -E UNICODE -A trust > $TD/initdb.out
		cat <<- EOF >> $TD/$PGVER/postgresql.conf
			    unix_socket_directories = '$TD'
			    listen_addresses = ''
			    shared_buffers = 12MB
			    fsync = off
			    synchronous_commit = off
			    full_page_writes = off
			    log_min_duration_statement = 0
			    log_connections = on
			    log_disconnections = on
		EOF
		touch $TD/NEW
		echo $TD
		;;
	start)
		# 1. Find a temporary database directory owned by the current user
		# 2. Create a new datadir if nothing was found
		# 3. Launch a background task to create a datadir for future invocations
		if [ -z $TD ]; then
			for d in $(ls -d ${SYSTMP:-/tmp}/ephemeralpg.*/$PGVER 2> /dev/null); do
				td=$(dirname "$d")
				test -O $td/NEW && rm $td/NEW 2> /dev/null && {
					TD=$td
					break
				}
			done
			[ -z $TD ] && {
				TD=$($0 initdb)
				rm $TD/NEW
			}
			nice -n 19 $0 initdb > /dev/null &
		else
			[ -O $TD/$PGVER ] || TD=$($0 initdb -d $TD)
		fi
		if [ ${TIMEOUT:-1} -gt 0 ]; then
			nice -n 19 $0 $KEEP -w ${TIMEOUT:-60} -d $TD -p ${PGPORT:-5432} stop > $TD/stop.log 2>&1 &
		fi
		[ -n "$PGPORT" ] && OPTS="-c listen_addresses='*' -c port=$PGPORT"
		LOGFILE="$TD/$PGVER/postgres.log"
		LC_ALL=${LC_ALL:-C} pg_ctl -W -o "$OPTS $USER_OPTS" -s -D $TD/$PGVER -l $LOGFILE start
		PGHOST=$TD
		export PGPORT PGHOST
		if [ -n "$PGPORT" ]; then
			url="postgresql://$(whoami)@$LISTENTO:$PGPORT/test"
		else
			url="postgresql:///test?host=$(echo $PGHOST | sed 's:/:%2F:g')"
		fi
		for n in 1 2 3 4 5; do
			sleep 0.1
			createdb -E UNICODE test > /dev/null 2>&1 && break
		done
		[ $? != 0 ] && {
			tail >&2 $LOGFILE
			exit 1
		}
		[ -t 1 ] && echo "$url" || printf "%s" "$url"
		;;
	stop)
		[ -O $TD/$PGVER/postgresql.conf ] || {
			echo >&2 "Please specify a PostgreSQL data directory using -d"
			exit 1
		}
		[ "$KEEP" == "" ] && trap "rm -r $TD" EXIT
		PGHOST=$TD
		export PGHOST PGPORT
		q="SELECT count(*) FROM pg_stat_activity
		WHERE datname IS NOT NULL
		AND state IS NOT NULL;"
		until [ "${count:-2}" -lt "2" ]; do
			sleep ${TIMEOUT:-5}
			count=$(psql test --no-psqlrc -At -c "$q" || echo 0)
		done
		pg_ctl -W -D $TD/$PGVER stop
		sleep 1
		;;
	selftest)
		export SYSTMP=$(mktemp -d /tmp/ephemeralpg-selftest.XXXXXX)
		trap "rm -r $SYSTMP" EXIT
		printf "Running: "
		printf "initdb "
		dir=$($0 initdb)
		printf "start "
		url=$($0 -w 3 -o '-c log_temp_files=100' start)
		printf "psql "
		[ "$(psql --no-psqlrc -At -c 'select 5' $url)" == "5" ]
		printf "stop "
		sleep 10
		printf "verify "
		! [ -d dir ]
		echo
		echo "OK"
		;;
	*)
		usage
		;;
esac
